### PR TITLE
mingw-w64-binutils 2.27 (new formula)

### PR DESCRIPTION
**Pull request by fxcoudert**
Cross-compiling binutils for mingw-w64 (Windows) target, 32 and 64-bit

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

File: Formula/mingw-w64-binutils.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+                          "--target=x86_64-w64-mingw32",
+                          "--enable-targets=x86_64-w64-mingw32,i686-w64-mingw32",
+                          "--prefix=#{prefix}",
+                          "--with-sysroot=#{prefix}"
+    system "make"
+    system "make", "install"
+
+    # Info pages and localization files conflict with native tools
+    info.rmtree
+    (share/"locale").rmtree
```
Comment (woodruffw) [inline Formula/mingw-w64-binutils.rb]: This might be a problem for non-English/default locale users. Any way around this?

Comment (fxcoudert) [inline Formula/mingw-w64-binutils.rb]: In the spirit of "minor conflicts" (as described in the cookbook), this is what is also done by other developer tool packages in homebrew-core, like gdb (https://github.com/Homebrew/homebrew-core/blob/master/Formula/gdb.rb). I followed the practice.

Another option would be to leave the locale files in the keg, and not symlinking them into `/usr/local/share/locale`. I have not found how to specify that in the recipe.

**Comment by lnxg33k** [comment]
👍 We need this man.

**Comment by fxcoudert** [comment]
Can I get an approval? Or a response to the earlier comment? (and if deemed necessary a link to doc explaining how to keep the locale files in the keg and prevent homebrew from symlinking them into `/usr/local/share/locale`)

File: Formula/mingw-w64-binutils.rb (hunk 1 above)
Comment (woodruffw) [inline Formula/mingw-w64-binutils.rb]: Thanks for the clarification. This seems like the right practice here.

**Comment by woodruffw** [comment]
@fxcoudert Sorry, I've been a bit busy. I'll merge this today.

File: Formula/mingw-w64-binutils.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  end
+
+  test do
+    assert_match "x86_64-w64-mingw32", shell_output("#{bin}/x86_64-w64-mingw32-as --version")
```
Comment (woodruffw) [inline Formula/mingw-w64-binutils.rb]: Can we build and link a simple assembly file here, instead of just testing `--version`?

**Comment by fxcoudert** [comment]
@woodruffw tests added. Cannot add proper tests for `ld` because we don't have any runtime libraries to link at that stage.

(Once this is merged, I will submit a PR for mingw-w64 headers, runtime and compiler.)

**Comment by woodruffw** [comment]
Shouldn't it be possible to run `ld` against two object files? We don't need to actually link against a library, just execute `ld` in a way that provides a reasonably good guarantee of functionality. Skipping `i386` is fine too, I think a combined `as`/`ld` test for `x86_64` provides a good guarantee.

Thanks for adding the `as` test. You can go ahead and remove the old `--version` tests.

**Comment by fxcoudert** [comment]
I added some bare bones linker tests. They link without CRT so they won't produce actual working executables, but at least we can check the format produced.

Given that binutils are multitarget (and it has been known to break in the past), I would rather have the tests enabled for both 32 and 64-bit targets, so that we know the multitarget configuration was correct.

**Comment by woodruffw** [comment]
> Given that binutils are multitarget (and it has been known to break in the past), I would rather have the tests enabled for both 32 and 64-bit targets, so that we know the multitarget configuration was correct.

That makes sense. Thanks for the explanation!

**Comment by woodruffw** [comment]
Thanks for contributing @fxcoudert! This is a great addition to Homebrew :tada:

**Comment by MikeMcQuaid** [comment]
Let's hold off merging any more mingw stuff until the compiler is ready. It's unclear to me how this is useful without the compiler.

**Comment by fxcoudert** [comment]
This one was a dependency of the compiler. Next the CRT and compiler should go in together (their build systems are intertwined): https://github.com/Homebrew/homebrew-core/pull/9037

**Comment by MikeMcQuaid** [comment]
Personally I think they should all be in a single formula as they aren't particularly useful without each other.

**Comment by fxcoudert** [comment]
With separate binutils, we have the potential for cross-compilers other than GCC (or other versions of GCC)

**Comment by MikeMcQuaid** [comment]
https://en.wikipedia.org/wiki/You_aren't_gonna_need_it 😉

**Pull request closed without merging**